From 358a1c0c89d958ddbbf157f06e0d5a0365b0fd7b Mon Sep 17 00:00:00 2001
From: Philippe Gerum <rpm@xenomai.org>
Date: Sun, 13 Nov 2022 15:25:36 +0100
Subject: [PATCH 783/985] evl/monitor: fix event untracking race upon
 concurrent wakeups

In the following scenario, the former implementation would yield a
double-removal of an event from the tracking list of its gate:

threadA:             wait_monitor(E, G)
<break/timeout>
                     [1] __evl_wait_schedule(E)
threadB:
                     enter_monitor(G)
                        signal_event(E)
                     exit_monitor(G)
                        __untrack_event(E)
                     enter_monitor(G)
                        signal_event(E)
                     exit_monitor(G)
                        __untrack_event(E)
threadA:             [2] untrack_event(E)

IOW, between points [1] and [2] of the code path returning from a wait
state on an error condition, other thread(s) may signal the event
which has no waiter anymore, therefore __untrack_event() would attempt
to unlink that event from the tracking list of its gate at every
release, leading to a list corruption such as:

Nov 11 21:25:07 10.243.1.49 [ 3877.744180] list_del corruption, ffff888151798240->next is LIST_POISON1 (dead000000000100)
Nov 11 21:25:07 10.243.1.49 [ 3877.846773] WARNING: CPU: 8 PID: 16134 at lib/list_debug.c:55 __list_del_entry_valid+0xc5/0x120
Nov 11 21:25:08 10.243.1.49 [ 3877.847043] Call Trace:
Nov 11 21:25:08 10.243.1.49 [ 3877.847045]  <TASK>
Nov 11 21:25:08 10.243.1.49 [ 3877.847059]  __untrack_event.part.9+0x26/0xe0
Nov 11 21:25:08 10.243.1.49 [ 3877.847096]  monitor_oob_ioctl+0x1356/0x1490
Nov 11 21:25:08 10.243.1.49 [ 3877.847367]  EVL_ioctl+0x81/0xf0
Nov 11 21:25:08 10.243.1.49 [ 3877.847386]  do_oob_syscall+0x699/0x6b0
Nov 11 21:25:08 10.243.1.49 [ 3877.847440]  handle_oob_syscall+0x189/0x230

Fix this by ensuring one-time removal based on the value of the
event->gate backlink.

Signed-off-by: Philippe Gerum <rpm@xenomai.org>
---
 kernel/evl/monitor.c | 46 ++++++++++++++++++++------------------------
 1 file changed, 21 insertions(+), 25 deletions(-)

diff --git a/kernel/evl/monitor.c b/kernel/evl/monitor.c
index 66671a8bd..7ce9f3d69 100644
--- a/kernel/evl/monitor.c
+++ b/kernel/evl/monitor.c
@@ -122,34 +122,25 @@ void __evl_commit_monitor_ceiling(void)
 	curr->u_window->pp_pending = EVL_NO_HANDLE;
 }
 
-/* event->gate->lock and event->wait_queue.wchan.lock held, irqs off */
-static void __untrack_event(struct evl_monitor *event)
+/* gate->lock + event->wait_queue.wchan.lock held, irqs off */
+static void untrack_event(struct evl_monitor *event, struct evl_monitor *gate)
 {
+	assert_hard_lock(&gate->lock);
+	assert_hard_lock(&event->wait_queue.wchan.lock);
+
 	/*
 	 * If no more waiter is pending on this event, have the gate
 	 * stop tracking it.
 	 */
-	if (!evl_wait_active(&event->wait_queue)) {
+	if (event->gate == gate && !evl_wait_active(&event->wait_queue)) {
 		list_del(&event->next);
 		event->gate = NULL;
 		event->state->u.event.gate_offset = EVL_MONITOR_NOGATE;
 	}
 }
 
-static void untrack_event(struct evl_monitor *event)
-{
-	struct evl_monitor *gate = event->gate;
-	unsigned long flags;
-
-	raw_spin_lock_irqsave(&gate->lock, flags);
-	raw_spin_lock(&event->wait_queue.wchan.lock);
-	__untrack_event(event);
-	raw_spin_unlock(&event->wait_queue.wchan.lock);
-	raw_spin_unlock_irqrestore(&gate->lock, flags);
-}
-
-/* event->gate->lock and event->wait_queue.wchan.lock held, irqs off */
-static void wakeup_waiters(struct evl_monitor *event)
+/* gate->lock + event->wait_queue.wchan.lock held, irqs off */
+static void wakeup_waiters(struct evl_monitor *event, struct evl_monitor *gate)
 {
 	struct evl_monitor_state *state = event->state;
 	struct evl_thread *waiter, *n;
@@ -160,16 +151,16 @@ static void wakeup_waiters(struct evl_monitor *event)
 	/*
 	 * We are called upon exiting a gate which serializes access
 	 * to a signaled event. Unblock the thread(s) satisfied by the
-	 * signal, either all, some or only one of them, depending on
-	 * whether this is due to a broadcast, targeted or regular
-	 * notification.
+	 * signal, either all of them, a designated set or the first
+	 * waiter in line, depending on whether this is due to a
+	 * broadcast, targeted or regular notification.
 	 *
 	 * Precedence order for event delivery is as follows:
 	 * broadcast > targeted > regular.  This means that a
 	 * broadcast notification is considered first and applied if
 	 * detected. Otherwise, and in presence of a targeted wake up
-	 * request, only the target thread(s) are woken up. Otherwise,
-	 * the thread heading the wait queue is readied.
+	 * request, only target threads are resumed. Otherwise, the
+	 * thread heading the wait queue is readied.
 	 */
 	if (evl_wait_active(&event->wait_queue)) {
 		if (bcast) {
@@ -184,7 +175,7 @@ static void wakeup_waiters(struct evl_monitor *event)
 		} else {
 			evl_wake_up_head(&event->wait_queue);
 		}
-		__untrack_event(event);
+		untrack_event(event, gate);
 	} /* Otherwise, spurious wakeup (fine, might happen). */
 
 	state->flags &= ~(EVL_MONITOR_SIGNALED|
@@ -274,7 +265,7 @@ static int exit_monitor(struct evl_monitor *gate)
 		list_for_each_entry_safe(event, n, &gate->events, next) {
 			raw_spin_lock(&event->wait_queue.wchan.lock);
 			if (event->state->flags & EVL_MONITOR_SIGNALED)
-				wakeup_waiters(event);
+				wakeup_waiters(event, gate);
 			raw_spin_unlock(&event->wait_queue.wchan.lock);
 		}
 	}
@@ -545,7 +536,12 @@ static int wait_monitor(struct file *filp,
 	 */
 	ret = evl_wait_schedule(&event->wait_queue);
 	if (ret) {
-		untrack_event(event);
+		raw_spin_lock_irqsave(&gate->lock, flags);
+		raw_spin_lock(&event->wait_queue.wchan.lock);
+		untrack_event(event, gate);
+		raw_spin_unlock(&event->wait_queue.wchan.lock);
+		raw_spin_unlock_irqrestore(&gate->lock, flags);
+
 		/*
 		 * Disable syscall restart upon signal (only), user
 		 * receives -EINTR and a zero status in this case. If
-- 
2.30.2

